### PR TITLE
Add reductions(_:_:)

### PR DESCRIPTION
**Pull request by danielctull**
### Description

This addition allows the caller to get a sequence of the results of reduce applied to up to and including the element of a sequence.

This adds an implementation for reductions as discussed in issue #25.

### Detailed Design

To achieve this, a new sequence type called Reductions is added to the library.

For ease of use, a new method is added to sequence to provide the reductions:

```swift
extension Sequence {
  public func reductions<Result>(
    _ initial: Result,
    _ transform: @escaping (Result, Element) -> Result
) -> Reductions<Result, Self>
```

The implementation of `subscript(position:)` calls `reduce` each time, so there's likely some improvements to be made around the Collection conformance.

### Documentation Plan

As the `reductions(_:_:)` method is the main entry point, this is documented with an example of use. The `Reductions` sequence type itself is lightly documented, which I believe follows the trend of the current APIs in the library.

In the guide, I have also gathered some quotes to show why the name `reductions` was chosen over `scan`.

### Test Plan

Arrays and sequences are used to test each code path, calling `reductions(_:_:)` on each. The empty sequence is also tested to ensure it yields an empty sequence.

### Source Impact

This is a purely additive change.

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](../CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

### CODE REVIEW

**Comment by CTMacUser** [comment]
I think you should make this an eager/lazy pair of methods.  The eager version returns an `Array`, and may use a throwing function for the combining closure.  The lazy version uses a non-throwing, but escaping function instead for its combing closure.  (Some of pull requests I've recently made show how you can share code between versions.)

But the important part is that this facility **cannot** be a collection, only a sequence, since dereferencing imposes an O(*n*) time penalty.

**Comment by timvermeulen** [comment]
All similar functions in other languages that I'm aware of include the initial value in the output (except for the reactive ones). Maybe we should too?

>But the important part is that this facility **cannot** be a collection, only a sequence, since dereferencing imposes an O(n) time penalty.

Subscripting indeed needs to be done in constant time, but that doesn't mean that `Reductions` can't be a collection. If the index stores the base index as well as the accumulated value up until that point, you can satisfy all of `Collection`'s performance requirements. `BidirectionalCollection` is definitely off the table, though.

**Comment by danielctull** [comment]
> All similar functions in other languages that I'm aware of include the initial value in the output (except for the reactive ones). Maybe we should too?

I was under the impression that others don't include the initial value. I think C++ has the ability to do either with an inclusive and exclusive variants – but I'm not a C++ person, I was just going on what I could make of their documentation.

Most of my uses in stuff like the advent of code puzzles generally require the initial value though. In drafting this, I had another function which used `chain` to prepend the initial value. I would love if we had both options easily available, and am open to options with respect to name. One idea I had was `reductions(including:_:)` and `reductions(excluding:_:)`, but I wasn't sure if this was making the excluding case too verbose?

> > But the important part is that this facility **cannot** be a collection, only a sequence, since dereferencing imposes an O(n) time penalty.
> 
> Subscripting indeed needs to be done in constant time, but that doesn't mean that `Reductions` can't be a collection. If the index stores the base index as well as the accumulated value up until that point, you can satisfy all of `Collection`'s performance requirements. `BidirectionalCollection` is definitely off the table, though.

Yep, after @CTMacUser's comment I went and took a look at the implementation. I got stuck with `index(_:offsetBy:)` for negative offsets, it feels like this hits the same problem as `BidirectionalCollection` where the reductions up to the given index have to be built up again.

**Comment by timvermeulen** [comment]
>I was under the impression that others don't include the initial value. I think C++ has the ability to do either with an inclusive and exclusive variants – but I'm not a C++ person, I was just going on what I could make of their documentation.

From what I can see, the exclusive variant takes an explicit initial value which is included in the result, and the inclusive variant uses the first element of the iterator as the initial value, which is then also included in the result. The initial value is included in the result either way, and the functions I checked in other languages seem to behave in the same way.

>I would love if we had both options easily available

It seems reasonable to me to have a variant of `reductions` that takes no initial value, i.e.

```swift
[1, 2, 3].reductions(+) // [1, 3, 6]
```

`reduce` doesn't have such a variant, but that probably doesn't need to stop us here.

Speaking of reduce, should `reductions(into:_:)` be included too? I believe it could benefit from the same copy-on-write gains that `reduce(into:_:)` was added for, as long as the user doesn't keep the partial result around when the next one is computed.

>I got stuck with `index(_:offsetBy:)` for negative offsets, it feels like this hits the same problem as `BidirectionalCollection` where the reductions up to the given index have to be built up again.

Without `BidirectionalConformance`, `index(_:offsetBy:)` doesn't need to be able to handle negative distances. But since the elements are always going to applied one by one, there isn't much point to override this method in the first place because you won't be able to improve on the default implementation.

**Comment by CTMacUser** [comment]
The reactive version of `scan` from Apple's `Combine` library does not vend the initial seed, although neither the documentation nor included example make that clear.

**Comment by timvermeulen** [comment]
Apparently ReactiveX does not output the seed either, [except for RxJava](https://stackoverflow.com/questions/48282685/rxjava-scan-emits-first-item-other-rx-libraries-do-not-why) 😅 But I have yet to come across a non-reactive scan that does not output the seed.

**Comment by pyrtsa** [comment]
For a non-reactive stream, I don't think it makes sense to discard the seed because it's so easy to skip by calling `dropFirst()` on the result. And the seed value is good to keep because it is often useful.

Defining `reductions` to include its seed value also aligns with the way [`sequence(first:next:)`](https://developer.apple.com/documentation/swift/2015879-sequence) is defined in the standard library.

Off-topic: For reactive streams if the result is also a stream of events, it probably makes sense to have the result of `scan` to match the length, and emission times, of the stream of incoming events. For those, the instant when to emit the seed would otherwise be left a bit ambiguous. The situation would be different if instead of a stream of discrete events the resulting value would be a continuous-time defined value (sometimes known as a _behaviour_). In that case it would start off immediately at the seed value and change to successive values right after every incoming event of inputs. I think we should view the result of `reductions` like that: it's the cumulative values for which the input sequence is the number of changes (and thus one less in length).

**Comment by danielctull** [comment]
These latest commits improve the implementation of the `Collection` conformance, remove `BidirectionalCollection` conformance and add two eager variants: one that takes no initial value and another which outputs the initial value at the start of the returned array. I realise lazy versions of these should also be created as well, but would rather we made sure what we wanted from the API before changing that part too. 🙂

In its current state the API looks like the following:

``` swift
[1, 2, 3].reductions(excluding: 10, +) // [11, 13, 16]
[1, 2, 3].reductions(including: 10, +) // [10, 11, 13, 16]
[1, 2, 3].reductions(+) // [1, 3, 6]
```

I quite like the point that if reductions included the initial value by default and the first API was removed, it is easy to produce the functionality by calling `dropFirst()` on the result, keeping the reductions API cleaner. If this was the consensus, the eager API would be the following, with matching lazy versions.

``` swift
[1, 2, 3].reductions(10, +) // [10, 11, 13, 16]
[1, 2, 3].reductions(+) // [1, 3, 6]
```

**Comment by pyrtsa** [comment]
I agree this would be ideal:
```swift
[1, 2, 3].reductions(10, +) // [10, 11, 13, 16]
[1, 2, 3].reductions(+) // [1, 3, 6]
```

**Comment by timvermeulen** [comment]
And potentially:
```swift
[1, 2, 3].reductions(into: 10, +=) // [10, 11, 13, 16]
```

`reductions(_:_:)` can always just call `reductions(into:_:)`, but the variant without a seed does probably need different internals.

**Comment by danielctull** [comment]
> And potentially:
> 
> ```swift
> [1, 2, 3].reductions(into: 10, +=) // [10, 11, 13, 16]
> ```

Yup! I was trying to summarise the comments so far and I failed, but I agree this should be added too. 🙂

**Comment by pyrtsa** [comment]
While it might be nice for symmetry, I don't see the eager `reductions(into:_:)` as having such a strong point for existence as [`reduce(into:_:)` originally did](https://developer.apple.com/documentation/swift/dictionary/values/3127145-reduce#discussion), because its implementation is nevertheless going to make _N_ modified copies in addition to the original seed.

The lazy version might be good though (because iterating over it could benefit from copy-on-write), and maybe the eager one indeed just for symmetry.

**Comment by timvermeulen** [comment]
The eager one can still have ergonomic advantages! Just no performance ones 🙂

**Comment by CTMacUser** [comment]
Eager versions can also use throwing closures.

**Comment by odmir** [comment]
```swift
let a = [1, 2, 3, 4].lazy.prefix(2).map { $0 }
let b = [1, 2, 3, 4].lazy.reductions(0, +).map { $0 }

print(type(of: a))
// LazyMapSequence<Slice<LazySequence<Array<Int>>>, Int>
print(type(of: b))
// Array<Int>
```
The current implementation loses the "laziness" information requiring one to apply `.lazy` again to the result.

File: Sources/Algorithms/Reductions.swift [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  /// - Parameters:
+  ///   - initial: The value to use as the initial accumulating value.
+  ///   - transform: A closure that combines an accumulating value and
+  ///     an element of the sequence.
+  /// - Returns: A sequence of transformed elements.
+  public func reductions<Result>(
+    _ initial: Result,
+    _ transform: @escaping (Result, Element) -> Result
+  ) -> Reductions<Result, Self> {
+    Reductions(base: self, initial: initial, transform: transform)
```
Comment (odmir) [inline Sources/Algorithms/Reductions.swift:149]:
```suggestion
  ) -> LazySequence<Reductions<Result, Self>> {
    Reductions(base: self, initial: initial, transform: transform).lazy
```

This should do it.

**Comment by danielctull** [comment]
> The current implementation loses the "laziness" information requiring one to apply .lazy again to the result.

Thanks @odmir, the conditional conformances to LazySequenceProtocol and LazyCollectionProtocol was missing.

File: Sources/Algorithms/Reductions.swift [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  where Base: LazySequenceProtocol {}
+
+extension Collection {
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element] {
+
+    guard let first = first else { return [] }
+    return try dropFirst().reductions(first, transform)
+  }
```
Comment (timvermeulen) [inline Sources/Algorithms/Reductions.swift:94]: In order to add this to `Sequence` without having to duplicate any logic, you could take the iterator, call `next` on it to get the seed, and then turn the iterator back into a sequence using [`IteratorSequence`](https://developer.apple.com/documentation/swift/iteratorsequence) (which hopefully optimizes well).

**Comment by danielctull** [comment]
I'm writing the lazy version of reductions and hit a question with the empty case if you fine people could guide me. 🙂

Given that `reductions` now includes the initial seed value, does that mean should `EmptyCollection<Int>().reductions(10, +)` yield `[10]`? Or should it return an empty array?

**Comment by odmir** [comment]
> > The current implementation loses the "laziness" information requiring one to apply .lazy again to the result.
> 
> Thanks @odmir, the conditional conformances to LazySequenceProtocol and LazyCollectionProtocol was missing. Using your example now yields the following. 🙂
> 
> ```swift
> let a = [1, 2, 3, 4].lazy.prefix(2).map { $0 }
> let b = [1, 2, 3, 4].lazy.reductions(0, +).map { $0 }
> 
> print(type(of: a))
> // LazyMapSequence<Slice<LazySequence<Array<Int>>>, Int>
> print(type(of: b))
> // LazyMapSequence<Reductions<Int, LazySequence<Array<Int>>>, Int>
> ```

Of course that's the correct way of doing it, sorry for the noise! 🤦

**Comment by xwu** [comment]
> Given that `reductions` now includes the initial seed value, does that mean should `EmptyCollection<Int>().reductions(10, +)` yield `[10]`?

Yes, and this appears to be how it behaves in other languages too. The version that takes an initial value would always have one more element in the result than are in the input collection, and the version that doesn’t, wouldn’t.

**Comment by danielctull** [comment]
> Of course that's the correct way of doing it, sorry for the noise! 🤦

@odmir: No noise at all, I appreciate you raising the concern! :)

**Comment by danielctull** [comment]
Thanks for confirming that @xwu!

**Comment by benlings** [comment]
Would it be worth adding a `deprecated` alias `scan` for this (eg. like [this](https://github.com/apple/swift-system/blob/8ffa04c0a0592e6f4f9c30926dedd8fa1c5371f9/Sources/System/FileOperations.swift#L131)? I have no problem with the name `reductions`, but would probably not have found it and assumed it didn't exist.

**Comment by danielctull** [comment]
Wow, I just remembered that I had this in flight. I moved house and have been really busy with that so haven't had much time to pursue this stuff.

Does anyone have any thoughts about @benlings' thought to include `scan` as a deprecated function to allow for easier discovery? Personally, I think the library is small enough that `reductions` would stand out on its own, but I'm happy to add a deprecated `scan` function.

**Comment by jasdev** [comment]
> Does anyone have any thoughts about @benlings’ thought to include `scan` as a deprecated function to allow for easier discovery?

+1 from me — I actually stumbled upon this PR by searching around for “scan.”

Thanks for driving — and following up on — this PR, @danielctull! 🙏🏽

File: Sources/Algorithms/Reductions.swift [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+extension Sequence {
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element] {
+    var iterator = makeIterator()
+    guard let initial = iterator.next() else { return [] }
+    return try IteratorSequence(iterator).reductions(initial, transform)
+  }
+}
```
Comment (timvermeulen) [inline Sources/Algorithms/Reductions.swift:180]: These are missing documentation as well. Might also be worth pointing out the relation to the version of `reductions` that takes an initial result.

**Review by timvermeulen — CHANGES_REQUESTED** [review]
It's nice to see this getting close to the finish line! I've left some more feedback. It would also be great to mark public methods with `@inlinable` and internal methods with `@usableFromInline` like we do in the rest of the package.

Let's indeed also add a deprecated `scan` method for discoverability purposes — we do not currently have a rigorous policy in place on when we want to add deprecated methods with names from prior art, but in this case `scan` is such a widely known name for this operation that it's worthwhile doing this.

**Comment by danielctull** [comment]
Cheers for those @timvermeulen! Will look at them over the next day. 😊

File: Guides/Reductions.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element]
+}
+```
+
+### Complexity
+
+Calling these methods is O(_1_).
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Guides/Reductions.md]: Resolved with 714f46bc63984947a32c910d2afd003215fde1af.

File: Sources/Algorithms/Reductions.swift [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+extension InclusiveReductions: LazySequenceProtocol
+  where Base: LazySequenceProtocol {}
+
+extension InclusiveReductions: LazyCollectionProtocol
+  where Base: LazyCollectionProtocol {}
+
+// MARK: - Shared ReductionsIndex
+
+public struct ReductionsIndex<BaseIndex: Comparable, Result>: Comparable {
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Sources/Algorithms/Reductions.swift]: This is resolved with 5182363625df0bfaaa1db5be9d5ec35eaa699acc.

File: Guides/Reductions.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+print(Array(runningTotal))
+// prints [0, 1, 3, 6, 10, 15]
+
+let runningTotal = (1...5).reductions(+)
+print(Array(runningTotal))
+// prints [1, 3, 6, 10, 15]
+
+let runningMinimum = [3, 4, 2, 3, 1].reductions(min)
+print(Array(runningMinimum))
+// prints [3, 3, 2, 2, 1]
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Guides/Reductions.md]: Yes! Resolved with 84b6ddc.

**Comment by timvermeulen** [comment]
Hey @danielctull, this is to let you know we'll soon be landing this addition and making the minor changes necessary in order to ship this in the next release. 🚀 Thank you for all your work on this, and feel free to make any changes you still want to make in the meantime.

**Comment by danielctull** [comment]
Cheers for the prompt @timvermeulen, I actually have some time tonight to get the rest of the modifications and documentation changes made that you suggested if that's alright?

**Comment by timvermeulen** [comment]
Totally!

File: Sources/Algorithms/Reductions.swift [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  ///   - transform: A closure that combines the previously reduced result and
+  ///   the next element in the receiving sequence.
+  /// - Returns: A sequence of transformed elements.
+  public func reductions<Result>(
+    _ initial: Result,
+    _ transform: @escaping (Result, Element) -> Result
+  ) -> ExclusiveReductions<Result, Self> {
+    ExclusiveReductions(base: self, initial: initial, transform: transform)
+  }
+}
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Sources/Algorithms/Reductions.swift]: Resolved with 7464876.

File: Guides/Reductions.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  public func reductions<Result>(
+    _ initial: Result, 
+    _ transform: (Result, Element) throws -> Result
+  ) rethrows -> [Result]
+  
+  public func reductions(
+    _ transform: (Element, Element) throws -> Element
+  ) rethrows -> [Element]
+}
+```
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Guides/Reductions.md]: Resolved with 174efb8c167ef28dbd3ffa3e3583f4c4275fc7a5.

File: Guides/Reductions.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
+# Reductions
+
+[[Source](https://github.com/apple/swift-algorithms/blob/main/Sources/Algorithms/Reductions.swift) |
+ [Tests](https://github.com/apple/swift-algorithms/blob/main/Tests/SwiftAlgorithmsTests/ReductionsTests.swift)]
+
+Produces a sequence of values.
+
+This has the behaviour of reduce, but instead of returning the final result
+value, it returns the a sequence of the results returned from each element.
```
[the remark this answers is not in the document]
Comment (danielctull) [reply Guides/Reductions.md:9]: Resolved with 1e47f860564036ea7cb05c21c2095fe176dab835.

**Comment by danielctull** [comment]
The doc comments are there, but I know they need some re-working. I'm planning to look at these tomorrow morning British time. 🙂 

@timvermeulen: I'm personally not convinced about removing `.start` from the indexes because it seems to make it a little harder to reason about generally, happy to discuss this though.

**Comment by kylemacomber** [comment]
@swift-ci please test

**Comment by timvermeulen** [comment]
@danielctull The last changes look great! The `start` case thing is a small tweak that we don't need to concern ourselves with right now, we can always revisit it later. Let's land this if you have no more changes you wanted to make.

**Comment by danielctull** [comment]
@timvermeulen: Thanks for that. Agreed, I'm all done with the changes for now. :)

**Review by timvermeulen — APPROVED** [review]
